### PR TITLE
fix: remainingSupply is zero edge case

### PR DESCRIPTION
**Pull request by zhongeric**
This is a new edge case since we now use `remainingSupply` in accounting calculations post rollover change.

## Problem

`toTokensRoundingUp` in `_checkpointAtBlock` intentionally rounds up tokens cleared each block. Over many blocks this rounding can push `$totalClearedQ96_X7 >= TOTAL_SUPPLY_Q96_X7`, making `remainingSupplyQ96X7()` return zero while `cumulativeMps` has not yet reached `MPS`. Without a guard, subsequent checkpoints would continue executing the sell logic with zero remaining supply — inflating `$totalClearedQ96_X7` and `$currencyRaisedQ96_X7` past their true values ("phantom tokens").

## Fix

Added a `remainingSupplyQ96X7_ > 0` check inside `_checkpointAtBlock` that skips token/currency accounting when supply is exhausted. `cumulativeMps += deltaMps` is moved outside this guard so the auction schedule still completes on time. `cumulativeMpsPerPrice` is not advanced (no tokens to price), which means bids in the exhaustion window receive no additional tokens but continue spending currency — a protocol-favoring outcome.

## Tests

- `test_WhenRemainingSupplyIsZero_DoesNotUpdateAccounting` — fuzz test (2500 runs) that forces `remainingSupplyQ96X7 == 0` mid-auction and asserts `$totalClearedQ96_X7`, `$currencyRaisedQ96_X7`, and `cumulativeMpsPerPrice` are unchanged after running to completion, while `cumulativeMps` reaches `MPS`.
- `test_WhenSupplyExhausted_BidReceivesNoMoreTokensButSpendsMoreCurrency` — lifecycle fuzz test (888 runs) that submits real bids, exhausts supply, then uses `CheckpointAccountingLib.calculateFill` to verify a bid receives zero additional tokens after exhaustion while currency spent remains positive.

### CODE REVIEW

File: src/ContinuousClearingAuction.sol [hunk 1]
```diff
@@ -280,48 +283,54 @@ contract ContinuousClearingAuction is
...
-                        ConstantsLib.MPS - _checkpoint.cumulativeMps // guaranteed to be > 0 because deltaMps > 0
-                    );
-                    // Total change in currencyRaised = currency raised above clearing + currency raised at clearing
-                    currencyRaisedDeltaQ96X7 = currencyRaisedDeltaQ96X7.add(currencyRaisedAtClearingQ96X7);
-                    // Track cumulative currency raised exactly at this clearing price (used for partial exits)
-                    _checkpoint.currencyRaisedAtClearingPriceQ96_X7 =
-                        _checkpoint.currencyRaisedAtClearingPriceQ96_X7.add(currencyRaisedAtClearingQ96X7);
+            ValueX7 remainingSupplyQ96X7_ = remainingSupplyQ96X7();
+            // Only need to update currencyRaised and totalCleared if there is remaining supply
+            if (ValueX7.unwrap(remainingSupplyQ96X7_) > 0) {
```
Comment (mgretzke) [inline src/ContinuousClearingAuction.sol:268]: Could this already skip if remainingSupplyQ96 is < Q96?

File: plan.md
Comment (mgretzke) [inline plan.md:1]: Can be deleted

**Review by mgretzke — COMMENTED** [review]
Worth looking into the failing test. Otherwise solid edge case protection. 
Its not really fixing the underlying core issue, but feels fine for such a specific type of issue.

File: src/ContinuousClearingAuction.sol (hunk 1 above)
Comment (zhongeric) [reply src/ContinuousClearingAuction.sol:268]: hm, maybe? I hesitate to do that

**Comment by zhongeric** [comment]
> Worth looking into the failing test. Otherwise solid edge case protection. Its not really fixing the underlying core issue, but feels fine for such a specific type of issue.

Huh so interesting. I checked on main and that fuzz run also fails. Seems to be a set of inputs which isn't passing test setup assertions, not related to this PR. Will make a separate PR to address.

**Comment by zhongeric** [comment]
Fixed that test in this commit on dev: https://github.com/Uniswap/continuous-clearing-auction/pull/323/changes/39843c8d3b8281837ddc62a5a5d396a0c90f970b

File: src/ContinuousClearingAuction.sol (hunk 1 above)
Comment (mgretzke) [reply src/ContinuousClearingAuction.sol:268]: Yeah, I am hesitant as well, but technically, if sub 1 token exists, its basically the same as zero?

File: src/ContinuousClearingAuction.sol [hunk 2]
```diff
@@ -258,48 +261,54 @@ contract ContinuousClearingAuction is
...
                 }
-            }
 
-            // Convert currency to tokens at price, rounding up, and update global cleared tokens.
-            // Intentional round-up leaves a small amount of dust to sweep, ensuring cleared tokens never exceed TOTAL_SUPPLY
-            // even when using rounded-up clearing prices on tick boundaries.
-            uint256 tokensClearedQ96X7 = ValueX7.unwrap(currencyRaisedDeltaQ96X7).toTokensRoundingUp(clearingPriceQ96);
-            $totalClearedQ96X7 = $totalClearedQ96X7.add(ValueX7.wrap(tokensClearedQ96X7));
+                // Convert currency to tokens at price, rounding up, and update global cleared tokens.
+                // Intentional round-up leaves a small amount of dust to sweep, ensuring cleared tokens never exceed TOTAL_SUPPLY
```
Comment (hensha256) [inline src/ContinuousClearingAuction.sol]: > ensuring cleared tokens never exceed TOTAL_SUPPLY

From what we talked about on slack, this is not true

Comment (hensha256) [reply src/ContinuousClearingAuction.sol]: > leaves a small amount of dust to sweep

is it actually sweepable? my understanding is that `sweepUnsoldTokens` sweeps:

```
unsoldTokens = remainingSupply();
```

however if a token is counted as "cleared" it then won't be sweepable right? So i think the dust isnt sweepable?

Comment (hensha256) [reply src/ContinuousClearingAuction.sol]: i feel like the sweeping thing would be non-trivial to fix? so probably not worth it

Comment (hensha256) [reply src/ContinuousClearingAuction.sol]: unless you added a sweep available at like `end + 1 year` where all unclaimed tokens can be swept

File: src/ContinuousClearingAuction.sol [hunk 3]
```diff
@@ -176,6 +176,9 @@ contract ContinuousClearingAuction is
         // Unwrap as we defer dividing by 1e7 by moving it to the LHS as multiplication
         uint256 remainingSupplyQ96X7_ = ValueX7.unwrap(remainingSupplyQ96X7());
         // If there are no more remaining supply or schedule, return the minimum clearing price
+        // Note: it is possible that because of rounding, remainingSupply can be zero even though
```
Comment (hensha256) [inline src/ContinuousClearingAuction.sol:179]: random edgecase... bids submitted after remainingSupply is 0, but before END_BLOCK is reached, receive zero tokens and pay full amount. Its really just a part of the fact that after remainingSupply is 0 bidders are paying to receive 0 tokens, but feels worse

Comment (hensha256) [reply src/ContinuousClearingAuction.sol:179]: could put a checlk in `_submitBid` that remaining supply is >0?

File: src/ContinuousClearingAuction.sol (hunk 2 above)
Comment (zhongeric) [reply src/ContinuousClearingAuction.sol]: Comment is wrong, it should be "round up leaves dust in the contract". It is not sweepable

Comment (zhongeric) [reply src/ContinuousClearingAuction.sol]: > unless you added a sweep available at like `end + 1 year` where all unclaimed tokens can be swept

Not sure how this is related

File: src/ContinuousClearingAuction.sol (hunk 3 above)
Comment (zhongeric) [reply src/ContinuousClearingAuction.sol:179]: Yea I could add it here:

```solidity
        // Revert if there are no more tokens to be sold
        if (_checkpoint.remainingMpsInAuction() == 0) revert AuctionSoldOut();
```

Comment (zhongeric) [reply src/ContinuousClearingAuction.sol:179]: Added in latest commit!

File: src/AuctionStorage.sol [hunk 4]
```diff
@@ -127,11 +127,15 @@ abstract contract AuctionStorage is IAuctionStorage {
 
     /// @inheritdoc IAuctionStorage
     function remainingSupplyQ96X7() public view returns (ValueX7) {
+        return _remainingSupplyQ96X7();
+    }
+
+    function _remainingSupplyQ96X7() internal view returns (ValueX7) {
         return TOTAL_SUPPLY_Q96X7.saturatingSub($totalClearedQ96X7);
```
Comment (hensha256) [inline src/AuctionStorage.sol]: this could now just be a regular `sub` - but maybe you've got it for safety?

File: src/libraries/ValueX7Lib.sol [hunk 5]
```diff
@@ -19,3 +19,16 @@ function add(ValueX7 a, ValueX7 b) pure returns (ValueX7) {
 function saturatingSub(ValueX7 a, ValueX7 b) pure returns (ValueX7) {
     return ValueX7.wrap(FixedPointMathLib.saturatingSub(ValueX7.unwrap(a), ValueX7.unwrap(b)));
 }
+
+/// @notice Return the minimum of two ValueX7 values
+/// @dev Branchless ternary evaluation for `a < b ? a : b`
+function min(ValueX7 a, ValueX7 b) pure returns (ValueX7) {
```
Comment (hensha256) [inline src/libraries/ValueX7Lib.sol:24]: does this save much vs just doing the `a < b ? a : b` directly?

File: test/btt/auction/checkpoint.t.sol [hunk 6]
```diff
@@ -98,4 +107,137 @@ contract CheckpointTest is BttBase {
...
+        // Fill at final checkpoint vs. at the start
+        (uint256 tokensAtEnd, uint256 currencyAtEnd) =
+            CheckpointAccountingLib.calculateFill(bid, cpFinal.cumulativeMpsPerPrice, cpFinal.cumulativeMps);
+
+        // Tokens received must not increase after supply is exhausted
+        assertEq(tokensAtEnd, tokensAtExhaustion, 'tokens should not increase after supply exhaustion');
+
+        // Currency spent must increase because the auction schedule continued
+        // (cumulativeMps advanced while cumulativeMpsPerPrice stayed flat)
+        assertGt(currencyAtEnd, 0, 'currency spent should be positive');
```
Comment (hensha256) [inline test/btt/auction/checkpoint.t.sol]: should this use `currencyAtExhaustion` which can be fetched from `calculateFill` above?

File: src/ContinuousClearingAuction.sol [hunk 7]
```diff
@@ -336,7 +345,9 @@ contract ContinuousClearingAuction is
         uint64 currentBlockNumberIsh = uint64(_getBlockNumberish());
         Checkpoint memory _checkpoint = _checkpointAtBlock(currentBlockNumberIsh);
         // Revert if there are no more tokens to be sold
-        if (_checkpoint.remainingMpsInAuction() == 0) revert AuctionSoldOut();
+        if (_checkpoint.remainingMpsInAuction() == 0 || ValueX7.unwrap(_remainingSupplyQ96X7()) == 0) {
```
Comment (hensha256) [inline src/ContinuousClearingAuction.sol:348]: are both of these needed? if `ValueX7.unwrap(_remainingSupplyQ96X7()) == 0` is it not guaranteed that `_checkpoint.remainingMpsInAuction() == 0`?

File: src/ContinuousClearingAuction.sol (hunk 2 above)
Comment (hensha256) [reply src/ContinuousClearingAuction.sol]: oh i just meant if you want the dust to be sweepable - which likely you dont care

File: src/ContinuousClearingAuction.sol (hunk 7 above)
Comment (hensha256) [reply src/ContinuousClearingAuction.sol:348]: claude tells me im wrong lol, ignore

File: src/libraries/ValueX7Lib.sol (hunk 5 above)
Comment (zhongeric) [reply src/libraries/ValueX7Lib.sol:24]: I believe so but I can double check

File: src/ContinuousClearingAuction.sol (hunk 7 above)
Comment (zhongeric) [reply src/ContinuousClearingAuction.sol:348]: Ah yeah this is the main point of this PR since tokensCleared is rounded up over time it can outpace the schedule

File: test/btt/auction/checkpoint.t.sol (hunk 6 above)
Comment (zhongeric) [reply test/btt/auction/checkpoint.t.sol]: great catch 🤦

File: src/AuctionStorage.sol (hunk 4 above)
Comment (zhongeric) [reply src/AuctionStorage.sol]: Will remove! not needed yea

**Review by hensha256 — APPROVED** [review]
